contracts/ClaimAndStake.sol

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/Multicall.sol";
+import "@uma/core/contracts/merkle-distributor/implementation/MerkleDistributorInterface.sol";
+
+interface AcceleratingDistributorInterface {
+    function stakeFor(
+        address beneficiary,
+        address stakedToken,
+        uint256 amount
+    ) external;
+}
+
+interface AcrossMerkleDistributorInterface is MerkleDistributorInterface {
+    function claimFor(MerkleDistributorInterface.Claim memory _claim) external;
+}
+
+/**
+ * @notice Allows claimer to claim tokens from AcrossMerkleDistributor and stake into AcceleratingDistributor
+ * atomically in a single transaction. This intermediary contract also removes the need for claimer to approve
+ * AcceleratingDistributor to spend its staking tokens.
+ */
+
+contract ClaimAndStake is ReentrancyGuard, Ownable, Multicall {
+    using SafeERC20 for IERC20;
+
+    // Contract which rewards tokens to users that they can then stake.
+    AcrossMerkleDistributorInterface public merkleDistributor;
+
+    // Contract that user stakes claimed tokens into.
+    AcceleratingDistributorInterface public acceleratingDistributor;
+
+    /**************************************
+     *               EVENTS               *
+     **************************************/
+    event SetMerkleDistributor(AcrossMerkleDistributorInterface indexed newMerkleDistributor);
+    event SetAcceleratingDistributor(AcceleratingDistributorInterface indexed newAcceleratingDistributor);
+
+    constructor() {}
```

Comment:
Did you want to set the merkleDistributor and acceleratingDistributor contracts as immutable? IMO, this would reduce security surface area and have few (if any) downsides since this contract could always be redeployed with different parameters.

Thoughts?

Reply:
To be honest the main reason why I didn't do this is because I couldn't figure out how to get the `deploy` scripts to work with a MerkleDistributor contract address imported from `contracts-v2`. I'll re review

Reply:
Ok this was easy since we import `contracts-v2`